### DISCUSSION TITLE
Viscous flow over a base

### DISCUSSION THREAD

**Discussion by wehrad** [idaholab/moose#26187] [Q&A Getting Started]
Hi everyone,

I'm working on a mesh representing a viscous material and I'd like the latter to flow over the equivalent of a rock base/bed.

I thought of swapping all elements with the "bottom" boundary to a new subdomain ID to which I could give rock properties instead of the viscous ones and hence move the material/bed boundary one element up. But I couldn't find how to do so... 

The end goal is to implement a Coulomb model at the interface between the viscous material and a rock base, at the bottom of my current mesh. But maybe there is actually a better way to do it from the beginning than changing subdomain IDs? 

So far I've only found solutions for a friction law between two media that are directly described in the input mesh.

Thank you a lot in advance for your help!

**Comment by GiudGiud** [comment]
Hello

The only tool to change subdomain IDs dynamically currently is:
https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html
https://mooseframework.inl.gov/source/userobjects/CutElementSubdomainModifier.html (in XFEM module)

If you are thinking of stopping the simulation and restarting, you can use the mesh generator to change subdomain ids:
https://mooseframework.inl.gov/source/meshgenerators/ParsedSubdomainMeshGenerator.html

Guillaume

**Reply by wehrad** [reply]
Thank you a lot! I think both of those objects wouldn't work for me unfortunately as the bottom of my mesh is not a regular plan and hence can't be selected with a bounding box...

Zooming out, do you know if I could actually implement viscoucs flow over a hard bed with a Coulomb model at the bottom of my mesh without explicitly modeling the bed itself? 
For example, my mesh is in contact with water at the front and I implemented the influence of water as a pressure on the side, without implementing the water body itself.

**Reply by GiudGiud** [reply]
so if the delineation boundary is complicated, do you think you could represent it with a mesh-division object?
https://mooseframework.inl.gov/syntax/MeshDivisions/index.html

That s a pretty general system.

then we can make a new subdomainID changer based on the mesh-division. 

> For example, my mesh is in contact with water at the front and I implemented the influence of water as a pressure on the side, without implementing the water body itself.

Yes maybe using a MultiApp approach. Make the water body be simulated in another app and only send the data you need on a somewhat regular basis?

**Reply by wehrad** [reply]
I've added a simple version of the mesh here: https://github.com/AdrienWehrle/mastodon/blob/devel/meshes/test_bed.e As you can see, the last layer of elements is 500m thick and the "bottom" boundary is on the top face of this layer, which corresponds to my rock base. Next step is to assign it a different subdomain ID than the rest of the material in the input file so I can give it a very high viscosity. 

An example of input file I'm working with right now is here: https://github.com/AdrienWehrle/mastodon/blob/devel/inputs/viscous_deformation/boundary_coupling/mooseice_vd_flow_transient_simpg.i

Maybe that's easier for you to see if this is possible with existing objects or needs development?

**Reply by wehrad** [reply]
it looks like the opposite of this one https://mooseframework.inl.gov/source/meshgenerators/BreakBoundaryOnSubdomainGenerator.html would be perfect, i.e. something like "BreakSubdomainOnBoundaryGenerator"...

**Reply by GiudGiud** [reply]
So you need a single layer of element to have this different block ID? sort of a boundary layer?

**Reply by wehrad** [reply]
Exactly @GiudGiud !

**Reply by wehrad** [reply]
In this case, the bottom layer. Or maybe it's easier to stitch an additional one but I doubt it on such a complex delineation...

**Reply by GiudGiud** [reply]
@aprilnovak @eshemon  is a BoundaryLayerSubdomainIDGenerator (changes the subdomain ID near boundaries) on your radar by any chance?

Otherwise I can make it.

**Reply by eshemon** [reply]
No, we don't have that on the radar at this point. Tagging @miaoyinb for awareness if you go ahead and implement!

**Reply by wehrad** [reply]
Thanks a lot to you all! At the end for the moment I decided to subdivide my blocks at the mesh generation step, and then apply the variables, kernels etc... to each existing subdomain ID! Thank you a lot for your help!
